toolkit/tools/pkg/imagecustomizerlib/customizeuki.go

```diff
@@ -0,0 +1,476 @@
+	UkiBuildDir  = "UkiBuildDir"
+	UkiOutputDir = "EFI/Linux"
+)
+
+func prepareUki(buildDir string, uki *imagecustomizerapi.Uki, imageChroot *safechroot.Chroot) error {
+	var err error
+
+	if uki == nil {
+		return nil
+	}
+
+	logger.Log.Infof("Enabling UKI")
+
+	// Check UKI dependency packages.
+	err = validateUkiDependencies(imageChroot)
+	if err != nil {
+		return fmt.Errorf("failed to validate package dependencies for uki:\n%w", err)
+	}
+
+	// Create necessary directories for UKI.
+	err = createUkiDirectories(buildDir, imageChroot)
+	if err != nil {
+		return fmt.Errorf("failed to create UKI directories:\n%w", err)
+	}
+
+	// This code installs the systemd-boot bootloader into the EFI system partition (ESP).
+	// Note: When proper support for systemd-boot is implemented, the `bootctl install` command
+	// will likely be invoked as part of the `hardResetBootLoader()` function.
+	//
+	// The command being executed is:
+	//     bootctl install --no-variables
+	// This performs the following steps:
+	//   1. Creates the necessary directories in the ESP, such as:
+	//        - /boot/efi/EFI/systemd
+	//        - /boot/efi/loader
+	//        - /boot/efi/loader/entries
+	//   2. Copies the systemd bootloader binary from the host filesystem to the ESP:
+	//        - Copies /usr/lib/systemd/boot/efi/systemd-bootx64.efi to /boot/efi/EFI/systemd/systemd-bootx64.efi
+	//        - Copies /usr/lib/systemd/boot/efi/systemd-bootx64.efi to /boot/efi/EFI/BOOT/BOOTX64.EFI
+	//          (This second location serves as the fallback bootloader entry, adhering to UEFI conventions.)
+	//   3. Writes a random seed to /boot/efi/loader/random-seed. This is used by the bootloader to initialize randomness.
+	//
+	// The "--no-variables" flag ensures that the command does not modify UEFI NVRAM boot variables. Instead, it relies
+	// on the bootloader binaries being present in the ESP for booting.
+	err = imageChroot.UnsafeRun(func() error {
+		return shell.ExecuteLiveWithErr(1, "bootctl", "install", "--no-variables")
+	})
+	if err != nil {
+		return fmt.Errorf("failed to install systemd-boot:\n%w", err)
+	}
+
+	// Start mapping kernels and initramfs.
+	bootDir := filepath.Join(imageChroot.RootDir(), BootDir)
+
+	kernelToInitramfs, err := getKernelToInitramfsMap(bootDir, uki.Kernels)
+	if err != nil {
+		return err
+	}
+
+	// Copy UKI-specific files such as kernel, initramfs, and UKI stub file.
+	err = copyUkiFiles(buildDir, kernelToInitramfs, imageChroot)
+	if err != nil {
+		return fmt.Errorf("failed to copy UKI files:\n%w", err)
+	}
+
+	// Extract kernel command line arguments from grub.cfg.
+	grubCfgPath := filepath.Join(bootDir, GrubCfgDir)
+	cmdlineArgs, err := extractKernelCmdlineFromGrub(grubCfgPath)
+	if err != nil {
+		return fmt.Errorf("failed to extract kernel command line arguments:\n%w", err)
+	}
+
+	// Dump kernel command line arguments to a file in buildDir.
+	cmdlineFilePath := filepath.Join(buildDir, UkiBuildDir, "kernel-cmdline-args")
```

Comment:
If there are multiple kernels installed, in theory, each kernel might have its own set of args associated with it. So, perhaps this file should be a JSON file containing a map from kernel to args.

That being said, this code will likely change when systemd-boot is implemented. So, perhaps this code is fine for now.